Firestore/core/src/firebase/firestore/core/sync_engine.h

```diff
@@ -47,6 +46,44 @@ namespace firebase {
 namespace firestore {
 namespace core {
 
+/**
+ * Interface implemented by `EventManager` to handle notifications from
+ * `SyncEngine`.
+ */
+class QueryEventCallback {
+ public:
+  /** Handles a change in online state. */
+  virtual void HandleOnlineStateChange(model::OnlineState online_state) = 0;
+  /** Handles new view snapshots. */
+  virtual void OnViewSnapshots(std::vector<core::ViewSnapshot>&& snapshots) = 0;
+  /** Handles the failure of a query. */
+  virtual void OnError(const core::Query& query, util::Status error) = 0;
+
+  virtual ~QueryEventCallback() = default;
+};
+
+/**
+ * Interface implemented by `SyncEngine` receive requests from `EventManager`.
+ */
+class QueryEventSource {
```

Comment:
Sorry if we discussed this before and I forgot: what's the reason for this decoupling? It doesn't exist on other platforms.

Reply:
Done.